File: frontend/lib/core/utils/email_policy.dart
/// Email policy helpers — mirrors BE EmailPolicy.
///
/// Kakao users who did not consent to email sharing receive a placeholder
/// address in the form `{provider}_{providerId}@no-email.local`.
/// An empty string is also treated as "not registered".
///
/// All screens must use [isPlaceholderEmail] / [User.hasRegisteredEmail]
/// from this single location — never inline the check.
const String _placeholderSuffix = '@no-email.local';

/// Returns true when [email] is a placeholder or empty, meaning the user
/// has not registered a real email address.
bool isPlaceholderEmail(String email) {
  return email.isEmpty || email.endsWith(_placeholderSuffix);
}

File: backend/src/main/kotlin/com/budgetbook/auth/domain/EmailPolicy.kt
package com.budgetbook.auth.domain

/**
 * Central authority for placeholder email policy.
 * All email generation, detection, and validation must use these helpers.
 * Never define the placeholder domain elsewhere.
 */
object EmailPolicy {

    const val PLACEHOLDER_EMAIL_DOMAIN = "no-email.local"

    /**
     * Generates a deterministic placeholder email from provider and providerId.
     * Format: {provider_lowercase}_{providerId}@no-email.local
     */
    fun buildPlaceholderEmail(provider: AuthProvider, providerId: String): String =
        "${provider.name.lowercase()}_${providerId}@$PLACEHOLDER_EMAIL_DOMAIN"

    /**
     * Returns true if the email is a real, user-provided email address.
     * An email is NOT real if it is blank or ends with the placeholder domain.
     */
    fun isRealEmail(email: String): Boolean =
        email.isNotBlank() && !email.endsWith("@$PLACEHOLDER_EMAIL_DOMAIN")

    /**
     * Returns true if the email is a placeholder.
     */
    fun isPlaceholderEmail(email: String): Boolean = !isRealEmail(email)
}
